crates/aptos-api-tester/src/main.rs

```diff
@@ -0,0 +1,541 @@
+        EntryFunctionCall::CodePublishPackageTxn {
+            metadata_serialized: bcs::to_bytes(&metadata).expect("PackageMetadata has BCS"),
+            code: blobs.clone(),
+        }
+        .encode();
+
+    // create and submit transaction
+    let pending_txn =
+        build_and_submit_transaction(client, account, payload, TransactionOptions::default())
+            .await?;
+    client.wait_for_transaction(&pending_txn).await?;
+
+    let blob = match blobs.get(0) {
+        Some(bytecode) => bytecode.clone(),
+        None => return Err(anyhow!("error while getting bytecode from blobs")),
+    };
+
+    Ok(HexEncodedBytes::from(blob))
+}
+
+/// Helper function that interacts with the message module.
+async fn set_message(client: &Client, account: &mut LocalAccount, message: &str) -> Result<()> {
+    // create payload
+    let payload = TransactionPayload::EntryFunction(EntryFunction::new(
+        ModuleId::new(account.address(), ident_str!("message").to_owned()),
+        ident_str!("set_message").to_owned(),
+        vec![],
+        vec![bcs::to_bytes(message)?],
+    ));
+
+    // create and submit transaction
+    let pending_txn =
+        build_and_submit_transaction(client, account, payload, TransactionOptions::default())
+            .await?;
+    client.wait_for_transaction(&pending_txn).await?;
+
+    Ok(())
+}
+
+/// Helper function that gets back the result of the interaction.
+async fn get_message(client: &Client, address: AccountAddress) -> Option<String> {
+    let resource = match client
+        .get_account_resource(
+            address,
+            format!("{}::message::MessageHolder", address.to_hex_literal()).as_str(),
+        )
+        .await
+    {
+        Ok(response) => response.into_inner()?,
+        Err(_) => return None,
+    };
+
+    Some(resource.data.get("message")?.as_str()?.to_owned())
+}
+
+/// Tests module publishing and interaction. Checks that:
+///   - module data exists
+///   - can interact with module
+///   - resources reflect interaction
+async fn test_module(client: &Client, account: &mut LocalAccount) -> Result<(), TestFailure> {
+    // publish module
+    let blob = publish_module(client, account).await?;
+
+    // check module data
+    let response = client
+        .get_account_module(account.address(), "message")
+        .await?;
+
+    let expected_bytecode = &blob;
+    let actual_bytecode = &response.inner().bytecode;
+
+    if expected_bytecode != actual_bytecode {
+        return Err(TestFailure::Fail("wrong bytecode"));
+    }
+
+    // interact with module
+    let message = "test message";
+    set_message(client, account, message).await?;
+
+    // check that the message is sent
+    let expected_message = message.to_string();
+    let actual_message = match get_message(client, account.address()).await {
+        Some(message) => message,
+        None => {
+            return Err(TestFailure::Error(anyhow!(
+                "module interaction isn't reflected"
```

Comment:
Are these supposed to be generic types?  Do these messages end up in metrics or just logs?  If they're supposed to be typed messages that can be searched later, it may be easier to put them in constants.